### PR TITLE
Text deltas v2

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/services/text_input.dart [hunk 1]
```diff
@@ -1859,6 +1859,85 @@ class TextInput {
     return true;
   }
 
+  TextEditingDelta _inferDeltas(Map<String, dynamic> rawDeltas) {
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:1862]: This method seems like a relic from our old way of doing things. It seems strange to create a new json Map and then create the TextEditingDelta. Maybe this logic should go into the fromJSON constructor?

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -791,23 +791,125 @@ abstract class TextEditingDelta with TextEditingDeltaUtils {
...
-        return TextEditingDeltaDeletion.fromJSON(encoded);
-      case 'TextEditingDeltaType.replacement':
-        return TextEditingDeltaReplacement.fromJSON(encoded);
-      case 'TextEditingDeltaType.nonTextUpdate':
-        return TextEditingDeltaNonTextUpdate.fromJSON(encoded);
-      default:
-        // Default clause should not be reachable.
-        assert(false);
-        return TextEditingDeltaNonTextUpdate.fromJSON(encoded);
+    String oldText = encoded['oldText'] as String;
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:797]: Do we need any additional error checking on this json anywhere?

File: packages/flutter/lib/src/services/text_input.dart [hunk 3]
```diff
@@ -791,23 +791,125 @@ abstract class TextEditingDelta with TextEditingDeltaUtils {
...
+          start: start,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+
+    assert(false);
+    return TextEditingDeltaNonTextUpdate(
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:908]: If a user creates some new subclass of TextEditingDelta that they want to create based on some special case of the json, would it be easy for them to do that?  Like overriding fromJSON and calling super.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_input.dart:908]: It sounds like they would just override the fromJSON. First they would check for their new delta type, if they cannot the delta based on their new delta type, then they call super.fromJSON which will try to infer based on the original delta types.

File: packages/flutter/lib/src/services/text_input.dart (hunk 2 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_input.dart:797]: I don't believe so 'oldText', 'deltaStart', and 'deltaEnd', 'deltaText' will never be null coming from the engine.

**Pull request merged**